### ISSUE TITLE
Refactor to Constraints when lcobucci/jwt 4.0 is released

### ISSUE THREAD

**Issue by steverhoades** [steverhoades/oauth2-openid-connect-client#2]
Reference: #https://github.com/lcobucci/jwt/issues/72

**Comment by lcobucci** [comment]
Don't forget that v4.0 will require `php: >=7.0`

**Comment by steverhoades** [comment]
@lcobucci this is a good point as PHP 5.6 is not EOL until Jan 2019.

**Comment by lcobucci** [comment]
@steverhoades indeed. I saw that you don't have this requirement in this package so was just wanted to remind you about this. v3 will still be maintained (without new features) to keep the compatibility with PHP 5 :wink:

**Comment by fanglingsu** [comment]
@steverhoades PHP 7.2 is already EOL and the lcobucci/jwt 4.2.x already available. Are there any plans to run this also under PHP8?

**Comment by Levivb** [comment]
I'm currently in the process of upgrading this package. Although I intend to keep the structure and interface as much intact as possible I do admit due to the upgrade to php 7.4 (because `lcobucci/jwt` requires php 7.4 as minimum) quite some code will change and will make use of type hinting and a bit more strict expectations.

See:
https://github.com/Levivb/oauth2-openid-connect-client

Will submit a PR later

**Comment by Levivb** [comment]
Also note that I lack full comprehension of OpenID Connect. So I have to take a careful look at https://lcobucci-jwt.readthedocs.io/en/latest/upgrading/ to see what has to be changed all the while keeping the code working. At the moment implementing (almost?) full test coverage with `lcobucci/jwt:3.4` to ease the upgrade to `lcobucci/jwt:4.1`

**Comment by lcobucci** [comment]
@Levivb that sounds great. Please let me know if there's anything I can do to help you!

**Comment by steverhoades** [comment]
It's hard to believe this issue is 5 years old now. I apologize it hasn't been upgraded already. @Levivb your contribution will be greatly appreciated!

**Comment by Levivb** [comment]
Good to hear 😄 

Currently all classes are fully tested except for the behemoth `OpenIDConnectProvider::getAccessToken` test which is WIP. I will continue that test tomorrow or possibly next week (got several meetings and appointments tomorrow).

Will check code coverage as well to be sure nothing is accidentally untested.
After that, the `lcobucci/jwt`can be updated to `4.1` and necessary code changes applied.

I also applied PSR12 through [phpcs](https://packagist.org/packages/squizlabs/php_codesniffer), added [enlightn/security-checker](https://packagist.org/packages/enlightn/security-checker), [ergebnis/composer-normalize](https://packagist.org/packages/ergebnis/composer-normalize), [insolita/unused-scanner](https://packagist.org/packages/insolita/unused-scanner), [maglnet/composer-require-checker](https://packagist.org/packages/maglnet/composer-require-checker), [php-parallel-lint/php-parallel-lint](https://packagist.org/packages/php-parallel-lint/php-parallel-lint) and [roave/security-advisories](https://packagist.org/packages/roave/security-advisories) to bring this package up to date with todays standards, quality and security.

**Comment by markwalet** [comment]
When can we expect support for `lcobucci/jwt:^4.0`? I'm working on upgrading the php version of one of my apps. And this package is the bottleneck right now.

The reason I'm asking is because the PR is already a few months in WIP :)

**Comment by JurRutten** [comment]
+1
